File: tests/metrics/probabilistic_sharpe_ratio_test.py
from __future__ import annotations
import numpy as np
import pytest
from scipy.stats import kurtosis, norm, skew
from taflow.metrics.probabilistic_sharpe_ratio import ProbabilisticSharpeRatio
VECTORBT_0_28_5_COMMIT = '993ceca7116fc8e55f4cd3a36fe43d83dab62b27'

def _vectorbt_probabilistic_sharpe(returns: np.ndarray, periods_per_year: float, annual_risk_free_rate: float, annual_benchmark_sharpe_ratio: float) -> float | None:
    """Translate vectorbt's pinned DSR kernel with an explicit benchmark SR."""
    usable = returns[~np.isnan(returns)]
    if len(usable) < 4:
        return None
    period_rf = np.expm1(np.log1p(annual_risk_free_rate) / periods_per_year)
    excess = usable - period_rf
    standard_deviation = float(np.std(excess, ddof=1))
    if standard_deviation == 0.0:
        return None
    estimated_sharpe = float(np.mean(excess) / standard_deviation)
    sample_skewness = float(skew(excess, bias=False))
    sample_kurtosis = float(kurtosis(excess, fisher=False, bias=False))
    adjustment = 1.0 - sample_skewness * estimated_sharpe + (sample_kurtosis - 1.0) * estimated_sharpe ** 2 / 4.0
    if not np.isfinite(adjustment) or adjustment <= 0.0:
        return None
    benchmark = annual_benchmark_sharpe_ratio / np.sqrt(periods_per_year)
    statistic = (estimated_sharpe - benchmark) * np.sqrt(len(usable) - 1.0) / np.sqrt(adjustment)
    return float(norm.cdf(statistic))

@pytest.mark.parametrize(('returns', 'periods_per_year', 'annual_risk_free_rate', 'benchmark'), [(np.array([0.02, -0.01, 0.03, -0.025, 0.01, -0.04, 0.015]), 252.0, 0.0, 0.5), (np.linspace(-0.04, 0.05, 101), 52.0, -0.01, -0.2), (np.array([0.15, -0.1, -0.1, 0.25, -0.3, 0.1]), 12.0, 0.04, 1.2), (np.array([np.nan, 0.02, -0.03, np.nan, 0.01, 0.04]), 252.0, 0.03, 0.0), (np.random.default_rng(20260812).standard_t(5.0, 513) * 0.012, 365.0, 0.05, 0.8)])
def test_probabilistic_sharpe_matches_pinned_vectorbt_formula(returns: np.ndarray, periods_per_year: float, annual_risk_free_rate: float, benchmark: float) -> None:
    assert len(VECTORBT_0_28_5_COMMIT) == 40
    actual = ProbabilisticSharpeRatio(periods_per_year=periods_per_year, annual_risk_free_rate=annual_risk_free_rate, annual_benchmark_sharpe_ratio=benchmark).from_returns(returns).compute()
    expected = _vectorbt_probabilistic_sharpe(returns, periods_per_year, annual_risk_free_rate, benchmark)
    if expected is None:
        assert actual is None
    else:
        assert actual == pytest.approx(expected, rel=2e-11, abs=2e-13)

def test_probabilistic_sharpe_input_methods_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.1, -0.2, 0.05, -0.03, 0.08, -0.01])
    settings = {'periods_per_year': 12.0, 'annual_risk_free_rate': 0.03, 'annual_benchmark_sharpe_ratio': 0.4}
    expected = ProbabilisticSharpeRatio(**settings).from_returns(returns).compute()
    equity = 100.0 * np.r_[1.0, np.cumprod(1.0 + returns)]
    pnl = np.diff(equity)
    assert ProbabilisticSharpeRatio(**settings).from_equity(equity).compute() == pytest.approx(expected)
    assert ProbabilisticSharpeRatio(**settings).from_pnl(pnl, initial_capital=100.0).compute() == pytest.approx(expected)
    assert ProbabilisticSharpeRatio(**settings).from_log_returns(np.log1p(returns)).compute() == pytest.approx(expected)
    state = ProbabilisticSharpeRatio(**settings).from_returns([])
    for value in returns[:3]:
        assert state.append(value) is state and state.value is None
    assert state.extend(returns[3:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.compute() == state.compute()
    assert state.reset() is state and len(state) == 0
    assert state.extend(returns).compute() == pytest.approx(expected)

def test_probabilistic_sharpe_edges_and_validation() -> None:
    assert ProbabilisticSharpeRatio().from_returns([]).compute() is None
    assert ProbabilisticSharpeRatio().from_returns([0.01, -0.01, 0.02]).compute() is None
    assert ProbabilisticSharpeRatio().from_returns([0.01] * 8).compute() is None
    assert len(ProbabilisticSharpeRatio().from_returns([np.nan, 0.02, -0.01, 0.03, 0.01])) == 4
    with pytest.raises(ValueError):
        ProbabilisticSharpeRatio(nan_policy='raise').from_returns([np.nan])
    with pytest.raises(ValueError):
        ProbabilisticSharpeRatio().from_returns([np.inf])
    with pytest.raises(ValueError):
        ProbabilisticSharpeRatio().from_returns([-1.01])
    with pytest.raises(ValueError):
        ProbabilisticSharpeRatio(periods_per_year=0.0).from_returns([0.01] * 4)
    with pytest.raises(ValueError):
        ProbabilisticSharpeRatio(annual_benchmark_sharpe_ratio=np.inf).from_returns([0.01] * 4)

def test_probabilistic_sharpe_requires_semantic_input_method() -> None:
    unbound = ProbabilisticSharpeRatio()
    with pytest.raises(ValueError):
        unbound.append(0.01)

File: crates/taflow-metrics/src/metrics/probabilistic_sharpe_ratio_test.rs
use super::probabilistic_sharpe_ratio::ProbabilisticSharpeRatio;
use crate::{MetricInputKind, NanPolicy};

fn assert_close(actual: f64, expected: f64, tolerance: f64) {
    assert!(
        (actual - expected).abs() <= tolerance,
        "{actual} != {expected}"
    );
}

#[test]
fn matches_pinned_vectorbt_formula_with_bias_corrected_sample_moments() {
    // vectorbt 0.28.5 commit 993ceca7116fc8e55f4cd3a36fe43d83dab62b27,
    // returns/metrics.py::deflated_sharpe_ratio. TAFlow substitutes a caller-
    // supplied benchmark Sharpe for vectorbt's estimated maximum Sharpe and
    // freezes scipy.stats skew/kurtosis with bias=False. The expected value was
    // evaluated independently with SciPy 1.17.0 norm.cdf/skew/kurtosis.
    let returns = [0.02, -0.01, 0.03, -0.025, 0.01, -0.04, 0.015];
    let mut state = ProbabilisticSharpeRatio::new(252.0, 0.0, 0.5, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    assert_eq!(state.extend(&returns[..3]).unwrap(), None);
    state.extend(&returns[3..]).unwrap();
    assert_close(state.compute().unwrap(), 0.469_251_442_389_235_66, 2e-12);
    assert_eq!(state.len(), returns.len());
}

#[test]
fn semantic_input_modes_and_streaming_lifecycle_are_invariant() {
    let returns = [0.10, -0.20, 0.05, -0.03, 0.08, -0.01];
    let make = |kind| {
        let mut state = ProbabilisticSharpeRatio::new(12.0, 0.03, 0.4, NanPolicy::Omit).unwrap();
        match kind {
            MetricInputKind::Returns => state.from_returns(&[]).unwrap(),
            MetricInputKind::LogReturns => state.from_log_returns(&[]).unwrap(),
            MetricInputKind::Equity => state.from_equity(&[]).unwrap(),
            MetricInputKind::PeriodPnl { initial_capital } => {
                state.from_pnl(&[], initial_capital).unwrap()
            }
            _ => unreachable!(),
        };
        state
    };
    let mut direct = make(MetricInputKind::Returns);
    let expected = direct.extend(&returns).unwrap().unwrap();

    let mut equity = make(MetricInputKind::Equity);
    assert_close(
        equity
            .extend(&[100.0, 110.0, 88.0, 92.4, 89.628, 96.798_24, 95.830_257_6])
            .unwrap()
            .unwrap(),
        expected,
        2e-12,
    );
    let mut pnl = make(MetricInputKind::PeriodPnl {
        initial_capital: 100.0,
    });
    assert_close(
        pnl.extend(&[10.0, -22.0, 4.4, -2.772, 7.170_24, -0.967_982_4])
            .unwrap()
            .unwrap(),
        expected,
        2e-12,
    );
    let mut logarithmic = make(MetricInputKind::LogReturns);
    assert_close(
        logarithmic
            .extend(&returns.map(f64::ln_1p))
            .unwrap()
            .unwrap(),
        expected,
        2e-12,
    );

    direct.reset();
    assert!(direct.is_empty());
    assert_eq!(direct.value(), None);
    for &value in &returns[..3] {
        assert_eq!(direct.append(value).unwrap(), None);
    }
    direct.extend(&returns[3..]).unwrap();
    assert_close(direct.compute().unwrap(), expected, 2e-12);
}

#[test]
fn handles_small_samples_constant_data_and_missing_values() {
    let mut state = ProbabilisticSharpeRatio::new(252.0, 0.0, 0.0, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    state.extend(&[f64::NAN, 0.01, -0.02, 0.03]).unwrap();
    assert_eq!(state.len(), 3);
    assert_eq!(state.compute(), None);
    state.append(-0.01).unwrap();
    assert!(state.compute().is_some());

    state.reset();
    state.extend(&[0.01; 8]).unwrap();
    assert_eq!(state.compute(), None);

    let mut raise = ProbabilisticSharpeRatio::new(252.0, 0.0, 0.0, NanPolicy::Raise)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    assert!(raise.append(f64::NAN).is_err());
    assert!(raise.is_empty());
    assert!(state.append(f64::INFINITY).is_err());
    assert!(state.append(-1.01).is_err());
}

#[test]
fn validates_configuration_and_semantic_domain() {
    for periods in [0.0, -1.0, f64::NAN, f64::INFINITY] {
        assert!(
            ProbabilisticSharpeRatio::new(periods, 0.0, 0.0, NanPolicy::Omit)
                .and_then(|mut state| {
                    state.from_returns(&[])?;
                    Ok(state)
                })
                .is_err()
        );
    }
    for rate in [-1.0, f64::NAN, f64::INFINITY] {
        assert!(
            ProbabilisticSharpeRatio::new(252.0, rate, 0.0, NanPolicy::Omit)
                .and_then(|mut state| {
                    state.from_returns(&[])?;
                    Ok(state)
                })
                .is_err()
        );
    }
    for benchmark in [f64::NAN, f64::INFINITY, f64::NEG_INFINITY] {
        assert!(
            ProbabilisticSharpeRatio::new(252.0, 0.0, benchmark, NanPolicy::Omit)
                .and_then(|mut state| {
                    state.from_returns(&[])?;
                    Ok(state)
                })
                .is_err()
        );
    }
    for kind in [MetricInputKind::RawPnl, MetricInputKind::Trades] {
        let _ = kind;
        let mut state = ProbabilisticSharpeRatio::new(252.0, 0.0, 0.0, NanPolicy::Omit).unwrap();
        assert!(state.append(0.0).is_err());
    }
}
